pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -786,72 +817,90 @@ public TransactionState getTransactionState() {
-    if (level == null) {
-      return Connection.TRANSACTION_READ_COMMITTED; // Best guess.
+    if (this.isolationLevel == -1 || !cacheXactState) {
+      ResultSet rs = null;
+      try {
+        rs = execSQLQuery("show transaction_isolation"); // nb: no BEGIN triggered
+        if (rs.next()) {
+          String transaction_isolation = rs.getString(1);
+          if ("read committed".equals(transaction_isolation)) {
+            this.isolationLevel = Connection.TRANSACTION_READ_COMMITTED;
+          } else if ("serializable".equals(transaction_isolation)) {
+            this.isolationLevel = Connection.TRANSACTION_SERIALIZABLE;
+          } else if ("repeatable read".equals(transaction_isolation)) {
+            this.isolationLevel = Connection.TRANSACTION_REPEATABLE_READ;
+          } else if ("read uncommitted".equals(transaction_isolation)) {
+            this.isolationLevel = Connection.TRANSACTION_READ_UNCOMMITTED;
+          }
+          LOGGER.log(Level.FINE, "  getTransactionIsolation = {0}", transaction_isolation);
+        }
+      } finally {
+        if (rs != null) {
+          rs.close();
+        }
+      }
     }
 
-    level = level.toUpperCase(Locale.US);
-    if (level.equals("READ COMMITTED")) {
-      return Connection.TRANSACTION_READ_COMMITTED;
-    }
-    if (level.equals("READ UNCOMMITTED")) {
-      return Connection.TRANSACTION_READ_UNCOMMITTED;
-    }
-    if (level.equals("REPEATABLE READ")) {
-      return Connection.TRANSACTION_REPEATABLE_READ;
-    }
-    if (level.equals("SERIALIZABLE")) {
-      return Connection.TRANSACTION_SERIALIZABLE;
+    // Double check we have a valid isolation level
+    if (this.isolationLevel == -1) {
+      LOGGER.log(Level.WARNING, "Could not get transaction isolation level.");
+      throw new PSQLException(GT.tr("Could not get transaction isolation level."),
+          PSQLState.TRANSACTION_STATE_INVALID);
     }
 
-    return Connection.TRANSACTION_READ_COMMITTED; // Best guess.
+    return this.isolationLevel;
   }
 
+  @Override
   public void setTransactionIsolation(int level) throws SQLException {
     checkClosed();
 
+    if (level == Connection.TRANSACTION_NONE) {
+      throw new PSQLException(GT.tr("Transaction isolation level {0} not supported.", "TRANSACTION_NONE"),
+          PSQLState.INVALID_PARAMETER_VALUE);
+    } else if (level == this.isolationLevel && cacheXactState) {
+      return; // Same isolation level, not set it again.
+    }
+
     if (queryExecutor.getTransactionState() != TransactionState.IDLE) {
       throw new PSQLException(
           GT.tr("Cannot change transaction isolation level in the middle of a transaction."),
           PSQLState.ACTIVE_SQL_TRANSACTION);
     }
 
-    String isolationLevelName = getIsolationLevelName(level);
+    String isolationLevelName;
+    switch (level) {
+      case Connection.TRANSACTION_READ_COMMITTED:
+        isolationLevelName = "READ COMMITTED";
```

Comment:
Is there a reason to move code around?

Reply:
Have a clean code?

Reply:
Typically we try to keep functional and style changes apart.
Frankly speaking, I do find "factored out" `getIsolationLevelName` easier to read.